### ISSUE TITLE
Should augmentations allow appending to the `implements` clause for extension types?

### ISSUE THREAD

**Issue by jakemac53** [dart-lang/language#4136]
We do not allow altering the representation type, and require the clause for that to be present, but we don't mention the `implements` clause.

cc @dart-lang/language-team

**Comment by lrhn** [comment]
I assume this means ~~extension~~ ~~mixin~~ types ... nope, I can't figure it out.

There is no declaration that has both an `on` clause and a `with` clause.  The only declaration which has a `with` clause is `class` and `enum` declarations, which cannot have an `on` clause.

The full grammars of the declarations are something like:
```dart
extension type const Name<X1, X2>(RepType id) implements I1, I2 { ... }
extension Name<X1, X2> on T { ... }
mixin Name<X1, X2> on T1, T2 implements I1, I2 { ... }
abstract base class Name<X1, X2> extends T with T1, T2 implements I1, I2 {}
abstract base mixin class Name<X1, X2> implements I1, I2 {}
enum Name<X1, X2> with T1, T2 implements I1, I2 {}
```

**Comment by jakemac53** [comment]
I just wrote it wrong, I meant the `implements` clause :)

**Comment by eernstg** [comment]
I think it should not be a problem to allow it. A longer `implements` clause would add to the available subtype relationships, but I don't think it will undo any existing assumptions for that declaration. Of course, there could be issues that correspond to the ones we can have with classes (like the non-existence of a combined member signature for a member, or a clash that can't be resolved like `int get foo` in one superinterface and `int f()` in another).

**Comment by jakemac53** [comment]
Yeah, I think it is mostly a philosophical question. We choose to for instance not allow filling in the on clause for extensions, or the representation type clause for extension types.

**Comment by eernstg** [comment]
Right, but adding an `implements` type to an extension type seems "no worse" than adding an `implements` type to a class, I think.

**Comment by jakemac53** [comment]
Yeah I am inclined to allow it 👍

**Comment by lrhn** [comment]
Agree. I'd prefer that `implements` behaves the same everywhere. It's stricty worse to be able to add interfaces to classes, mixins and enums, but not extension types.

**Comment by munificent** [comment]
As of #4357, the proposal allows this in the grammar, so I think we're good.
